### ISSUE TITLE
k_statistic / k_statistic_bootstrap: severe size distortion with within-cell-demeaned moments + NaN-mask bug in bootstrap

### ISSUE THREAD

**Issue by ligon** [ligon/ManifoldGMM#51]
## Summary

While running a Monte Carlo size check on the Kleibergen K-statistic in a real research project (cluster-robust GMM under the Athey-Imbens 2017 fine-stratification design, with 90 (stratum × arm) cells and a moment vector demeaned within (period, asset) cells), I found two issues:

1. **`k_statistic` is catastrophically mis-sized** in this setup. At nominal 5% level, the empirical rejection rate is ~95% (under H0 with σ_true=0.8). The K mean ≈ 12 vs χ²(1) mean = 1; K 95th percentile ≈ 21 vs χ²(1) 95% = 3.84.

2. **`k_statistic_bootstrap` doesn't fix this AND has a NaN-handling bug**: it uses `g_matrix.mean(axis=0)` at `bootstrap.py:1240`, not NaN-aware. With moment functions that emit NaN for non-applicable observations (e.g., delta-masked asset-pricing moments), `K_observed` comes back NaN and the bootstrap p-values are meaningless (= 1/(B+1) by the always-False NaN comparison).

The standalone `k_statistic` (asymptotic path) does handle NaN correctly via `restriction.g_bar`, so issue (2) is specific to the bootstrap path.

## Issue 1: size distortion of asymptotic K

Repro (paraphrased; full code in [Seasonality repo](https://github.com/ligon/Seasonality) `scripts/euler_mc_size.py`):

```python
# Generate datasets under H0 at sigma_true=0.8:
#   m_{i,t+1} log-normal AR(1) within i, iid across i
#   independent of (R, delta, z=cash dummy from random stratum-level
#   group assignment), so E[u·z]=0 holds by construction.
# Moment: u^j_{it} = (m_{i,t+1}^σ - mbar^j_{t+1}) · R^j_{t+1} · δ^j_{i,t-1}
# where mbar^j_t is the SAMPLE mean of m^σ within (t, j-holders).
# Cluster level: stratum × arm (G=90 cells, ~30 rows each).

for s in range(20):
    obs_s = dgp.draw()
    restriction = MomentRestriction(
        g=g, data=obs_s, clusters=stratum_arm_codes,
        manifold=Manifold.from_pymanopt(Positive(1, 1)),
        backend='jax', parameter_labels=('sigma',),
    )
    gmm = GMM(restriction=restriction, initial_point=jnp.array([[1.0]]))
    result = gmm.estimate(...)
    kr = result.k_statistic(theta_0=jnp.array([[0.8]]))
    # Record kr.K, kr.p_K
```

Results:

| stat | observed (S=20) | χ²(1) reference |
|---|---|---|
| K mean | 12.4 | 1.0 |
| K median | 11.5 | 0.45 |
| K 95% quantile | 20.8 | 3.84 |
| Empirical rejection rate at nominal 5% | **0.95** | 0.05 |
| Empirical rejection rate at nominal 1% | **0.90** | 0.01 |

The J-statistic in the same setup is mildly over-sized (~10% at nominal 5%) — not great but not catastrophic. K is far worse.

**Suspected root cause:** the moment function demeans `m^σ` within each `(t, j-holders)` cell using the sample mean. That makes `mbar^j_t` a *generated regressor* whose σ-derivative is missing from the K-stat's Jacobian D. Standard K-stat asymptotic theory (Kleibergen 2005) assumes the moment is an average of iid `g_i(θ)`; the within-cell sample-mean subtraction introduces dependence that breaks the χ²(p) reference. Equivalent issue would appear for any moment that includes a sample mean as a centering term (e.g., demeaned Euler equations, certain asset-pricing tests, panel within-transformations).

The fact that the moment is *constructed* with sample-mean demeaning isn't visible to the K-stat — it just sees `restriction.gi(θ)` returning per-row moment values. So K's implementation can't know about the generated-regressor structure.

**Possible fixes:**
- Document this limitation prominently in `k_statistic`'s docstring: "Asymptotic χ²(p) reference assumes the moment vector is an iid average; moments involving generated regressors (sample-mean demeaning, etc.) may produce a non-χ² null distribution."
- Or: detect and warn when the moment function appears to include sample-mean subtraction (hard heuristic).
- Or: provide a proper finite-sample diagnostic API (the cluster-wild bootstrap doesn't fix this — see issue (2) — because it holds Ω̂ and D fixed across replications).

## Issue 2: NaN bug in `k_statistic_bootstrap`

At `bootstrap.py:1240`:

```python
g_mean = g_matrix.mean(axis=0)  # shape (ell,)
g_bar = sqrt_N * g_mean
```

`g_matrix` here is `restriction.moment_contributions(eval_point)`, which can contain NaN entries (rows where the moment is undefined for that column, e.g., non-asset-holders in an asset-pricing test). `np.mean` propagates NaN: `g_mean` is NaN for any column with at least one NaN row → `g_bar` is NaN → `K_observed` is NaN → `K_bootstrap >= K_observed` is always False → `p_K_bootstrap = 1/(B+1)` regardless of the data.

The standalone `k_statistic` doesn't have this issue: it uses `restriction.g_bar(eval_point)` which is NaN-aware (delegates to `_mean` which uses `nanmean`).

**Fix:** use `restriction.g_bar(eval_point)` instead of `sqrt(N) * g_matrix.mean(axis=0)`. Or use `np.nanmean` and the per-column non-NaN count for `sqrt(N_k)` scaling (matching the convention in `MomentRestriction._mean` / `_count`).

A local NaN-aware workaround is at https://github.com/ligon/Seasonality/blob/main/scripts/euler_kstat_bootstrap_nanaware.py (reproduces the same K_observed as `k_statistic` plus a sensible bootstrap reference distribution).

## Versions

- ManifoldGMM: 0.4.0a (the local-path editable install)
- DGP_Protocol: 0.1.0a0 (post-#1, JAX-aware moment_covariance)
- Python 3.11

## Cross-references

- Filed alongside #5 (iid `omega_hat` NaN bug) in the consumer-side Seasonality repo's task tracker.
- Related: the v2 GMM auto-synthesized restriction also doesn't carry `cluster_ids` from `dgp.sampling` (separate gap; we work around by using v1 path `MomentRestriction(clusters=...)` for K-stat).

🤖 Filed via Claude Code while running cluster-robust GMM for an asset-pricing Euler test (Seasonality project).

**Comment by ligon** [comment]
## Update: MomentWildBootstrap J output isn't a null reference for the J test

Tangential observation while working around #51 / #52 in the same project. Posting here as a note rather than a separate issue because it's documentation-shaped rather than a bug.

`MomentWildBootstrap` re-fits `theta_star` per replicate and returns `criterion_value` = J at `theta_star`. This gives a useful object — the sampling distribution of J at refitted theta under wild resampling — but it's **not the same as** the null reference distribution most users want for an overid J test.

For a typical Hansen-J test "is the model misspecified at theta_hat?", the natural wild-bootstrap reference is:

1. Hold theta_hat fixed; compute centered moment contributions `u_i = g_i(theta_hat) - g_bar(theta_hat)`.
2. For each replicate: weight u_i by Rademacher cluster signs, evaluate `J* = weighted_g_bar' W weighted_g_bar` at `theta_hat` (W also at theta_hat).
3. Bootstrap p-value: `(1 + #{J* >= J_observed}) / (1 + B)`.

With re-fitting, `J*` distribution is centered near zero (theta_star is the J-minimizer on the weighted data), which makes a near-zero J_observed look anomalously large. In our setting J_observed = 7.07 vs J*_mean = 0.14 → bootstrap p ≈ 0.05, but this doesn't mean the model is misspecified — it means the bootstrap reference is a different statistic than we want.

Worth documenting in MomentWildBootstrap's docstring (something like: "criterion_value is the refit-J under wild resampling; for a no-refit null reference, use `gmm.k_statistic_bootstrap(theta_0=theta_hat)` and look at `J_bootstrap` instead"). Or expose a no-refit `J_bootstrap_null` method.

The k_statistic_bootstrap path *does* compute the right J* (no refit; see bootstrap.py:1277 `J_bootstrap = einsum(...)`), just that the surrounding scaffolding is K-focused. Pulling J* out of that path is what we'll do as the workaround.
